source/_posts/Docker-Swarm.md

```diff
@@ -0,0 +1,240 @@
+---
+title: Docker Swarm, Traefik, Let's Encrypt en Portainer
+date: 2019-11-03 12:00:00
+tags:
+- Docker swarm
+- Let's Encrypt
+- Traefik
+- Portainer
+---
+
+Case omschrijving 
+---
+Niet lang geleden heb ik besloten om me eens goed te gaan verdiepen in Docker. Je hoort het overal en het wordt steeds populairder en ik wilde eens gaan kijken wat hier nu zo vet aan is. Ik ben begonnen om me te certificeren, ik heb een cursus gevolgd en mijn examen met succes behaald. Ik vind het altijd belangrijk om te weten hoe iets in elkaar zit en waarom het zo in elkaar zit. Mijn idee is om een of meerdere productie applicaties op een Docker Swarm cluster te gaan draaien. In deze blog ga ik laten zien hoe ik een goede infrastructuur bouw om de applicaties te hosten. 
+
+De omgeving moets aan de volgende voorwaarden voldoen:
+
+- Er moeten meerdere web applicaties op kunnen draaien welke op poort 80 en 443 benaderd kunnen worden.
+- Alle web applicaties moeten draaien op SSL, dit moet me zo min mogelijk werk kosten.
+- De oplossing moet schaalbaar zijn en bestand zijn tegen het uitvallen en/of offline gaan van servers.
+- Ik wil de oplossing kunnen hosten bij een Cloud provider, Hosting provider of lokaal om te kunnen testen.
+- Het beheren van de containers moet gemakkelijk en overzichtelijk zijn.
+- Aangezien ik een echte Hollander ben moeten de kosten ook een beetje beperkt blijven. 
+
+De oplossing welke ik gemaakt heb is een combinatie van Docker Swarm, Traefik, Let's Encrypt en Portainer.
+
+Wat is Docker Swarm
+---
+Docker Swarm is een tool waarmee je Docker-containers kunt beheren en schalen. Als je Docker installeert dan krijg je daar meteen Swarm bij. Docker Swarm is een standaard product van Docker wat bij elke installatie van Docker wordt meegeleverd. Met Docker Swarm kun je een cluster bouwen van verschillende virtuele machines deze worden hierna nodes genoemd. Hierop kunnen de Docker-containers worden gedeployed als een stack\*\* of een service\*. Als je een cluster wilt hebben wat een hoge beschikbaarheid heeft wordt aangeraden om minstens 3 manager nodes te hebben. Ik ga hier niet uitleggen hoe je een Docker Swarm cluster moet bouwen, hier zijn hele goede handleidingen voor bijv. in de Docker documentatie https://docs.docker.com/engine/swarm/swarm-tutorial/create-swarm/ 
+
+<img src="/images/Docker-Swarm-overview.png" />
+Een overzicht hoe het cluster eruit ziet met 3 managers en 2 workers.
+
+Omdat ik Docker Swarm gebruik kan ik de volgende punten afvinken van mijn lijstje:
```

Comment:
Oh dit gaat snel, je hebt 3 nodes en meteen is iets schaalbaar en bestand tegen uitvallen?
Wat gebeurt er als 1 van je VM's kapot gaat?
Tot hoever kan je schalen? En wat als je resources van je VM's op zijn? Hoe provision je dan meer?